### PR TITLE
Calculate the cart total

### PR DESCRIPTION
**Pull request by philipr-za**
Closes https://github.com/big-neon/bn-api/issues/272

Added a method to the Order DB entity to calculate the total price of an order.
To do this I had to flesh out adding tickets to the order by determining the current ticket price, current fee schedule and creating a ticket order item and adding a fee order item.

### CODE REVIEW

File: db/src/models/mod.rs [hunk 1]
```diff
@@ -1,3 +1,4 @@
+#![allow(proc_macro_derive_resolution_fallback)]
```
Comment (mikethetike) [inline db/src/models/mod.rs]: Is this needed? Perhaps put a comment saying why it is here

File: db/src/models/orders.rs [hunk 2]
```diff
@@ -78,7 +81,56 @@ impl Order {
...
+        let ticket_pricing = TicketPricing::get_current_ticket_pricing(ticket_type_id, conn)?;
+
+        let organization = Organization::find(
+            Event::find(TicketType::find(ticket_type_id, conn)?.event_id, conn)?.organization_id,
+            conn,
+        )?;
+
+        if organization.fee_schedule_id.is_none() {
+            return Err(DatabaseError::new(
+                ErrorCode::NoResults,
```
Comment (mikethetike) [inline db/src/models/orders.rs]: Use a different error type here, otherwise it might become a 404 on the API

File: db/src/models/orders.rs [hunk 3]
```diff
@@ -78,7 +81,56 @@ impl Order {
...
+            ));
+        }
+
+        let fee_schedule_range = FeeSchedule::find(organization.fee_schedule_id.unwrap(), conn)?
+            .get_range(ticket_pricing.price_in_cents, conn)?;
+
+        //I would like to have used an .unwrap_or_else but I dont think I can return this error value from inside the closure
+        if fee_schedule_range.is_none() {
+            return Err(DatabaseError::new(
+                ErrorCode::NoResults,
```
Comment (mikethetike) [inline db/src/models/orders.rs]: Again, use a different ErrorCode. Preferrably one that gets mapped to HTTP 500.

File: db/src/models/organizations.rs [hunk 4]
```diff
@@ -286,14 +286,17 @@ impl Organization {
         &self,
         fee_schedule: &FeeSchedule,
         conn: &PgConnection,
-    ) -> Result<(), DatabaseError> {
+    ) -> Result<Organization, DatabaseError> {
+        let attributes = OrganizationEditableAttributes {
+            fee_schedule_id: Some(fee_schedule.id),
```
Comment (mikethetike) [inline db/src/models/organizations.rs]: Thinking about it, the fee schedule should always be set on an organization. We should set it on creation. If needs be, we could create a Zero Fees schedule if there need to be organizations with no fees. That would make the checking code a lot simpler. Perhaps create a seperate ticket to do this.

File: db/src/models/orders.rs (hunk 3 above)
Comment (philipr-za) [reply db/src/models/orders.rs]: I will do this, in the mean time I wanted to ask if you could think of a more Rustacean method for this check and the subsequent unwrap? I felt like I was missing an easy solution here.

File: db/src/models/mod.rs (hunk 1 above)
Comment (philipr-za) [reply db/src/models/mod.rs]: Will do, this is what suppresses those new warnings in our code.

File: db/src/models/orders.rs (hunk 2 above)
Comment (philipr-za) [reply db/src/models/orders.rs]: Changed to Business process error

File: db/src/models/orders.rs (hunk 3 above)
Comment (philipr-za) [reply db/src/models/orders.rs]: Changing to BusinessProcess Error

File: db/src/models/organizations.rs (hunk 4 above)
Comment (Jasonvdb) [reply db/src/models/organizations.rs]: I'm only allowing the fee schedule to be set once, then I need to allow it to be updated?

Comment (philipr-za) [reply db/src/models/organizations.rs]: So we have just made a change here that means you have to provide a fee_schedule_id when you create the Org. We dont want to allow an org to be created without a fee_schedule.

Comment (Jasonvdb) [reply db/src/models/organizations.rs]: How would I create a fee schedule without having an org first?

Comment (philipr-za) [reply db/src/models/organizations.rs]: Hmm that is a good point. I suppose I better include that as part of the Organzation create endpoint.

Comment (philipr-za) [reply db/src/models/organizations.rs]: Just for the record we are going to open another issue to discuss how we are handling the fee schedules.

To summarise the current discussion:
- We don't think that the fee_schedule_id field of the Org table should be nullable. There should ALWAYS be a fee schedule associated with an Org
- This means we need to create a fee schedule before creating an Org
- Current work-around is to create a default zero fee schedule as the Org is created which will then be over-written in the front-end.
- Will make an urgent issue to refactor the fee schedule database entities so this process flows better.

**Comment by philipr-za** [comment]
Closes https://github.com/big-neon/bn-api/issues/299

**Pull request merged**